### PR TITLE
[feat] Qwen 3 Omni MOE with EP support

### PR DESCRIPTION
**Pull request by ngquangtrung57**
## Motivation
Add Qwen 3 Omni MOE with EP support.

## Commit Message Convention

Please follow our standardized commit message format:

- `[feat]` - New features or functionality
- `[fix]` - Bug fixes
- `[docs]` - Documentation changes only
- `[style]` - Code style changes (formatting, missing semicolons, etc.)
- `[refactor]` - Code refactoring without changing functionality
- `[perf]` - Performance improvements
- `[test]` - Adding or updating tests
- `[chore]` - Maintenance tasks, dependency updates, etc.
- `[ci]` - CI/CD configuration changes

**Examples:**
- `[feat] add qwen omni iterable dataset support`
- `[fix] resolve bagel model configuration error`
- `[docs] update training guide with YAML examples`

See [CONTRIBUTING.md](../CONTRIBUTING.md) for more details.

## CI/CD Checks

Your PR will automatically run the following checks:

- **Linting**: Code formatting with `black` (line-length=120) and import sorting with `isort`
- Run `pre-commit run --all-files` locally to verify before pushing

## Checklist

- [x] Follow commit message convention (see above)
- [x] Run `pre-commit run --all-files` and ensure all checks pass
- [x] Format your code with `black` (line-length=120) and `isort`
- [x] Add unit tests for new functionality
- [x] Update documentation as needed, including docstrings or example tutorials
- [x] Ensure all CI/CD checks pass

### CODE REVIEW

**Comment by ngquangtrung57** [comment]
Currently testing on test dataset with EP=8 on 8 H100, the mfu is around 0.11 - 0.14. But I can only test ep cicd because without EP, using SP or without both cause OOM.

**Review by Luodian — APPROVED** [review]

File: src/lmms_engine/datasets/processor/__init__.py [hunk 1]
```diff
@@ -19,12 +20,12 @@
     "BaseQwen2_5_DataProcessor",
     "LLaVADataProcessor",
     "Qwen2_5OmniDataProcessor",
+    "Qwen3OmniMoeDataProcessor",
     "Qwen2_5_VLDataProcessor",
     "Qwen2VLDataProcessor",
     "WanVideoDataProcessor",
     "PureTextDataProcessor",
     "Qwen2DataProcessor",
-    "WanVideoDataProcessor",
```
Comment (kcz358) [inline src/lmms_engine/datasets/processor/__init__.py]: Is this one deleted accidentally?

File: src/lmms_engine/models/qwen3_omni_moe/qwen3_omni_moe_ops.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                hidden_states,
+                position_embeddings=position_embeddings,
+                attention_mask=attention_mask,
+                position_ids=position_ids,
+                past_key_values=past_key_values,
+                cache_position=cache_position,
+                cu_seq_lens=cu_seq_lens,
+                indices=indices,
+                **kwargs,
+            )
```
Comment (kcz358) [inline src/lmms_engine/models/qwen3_omni_moe/qwen3_omni_moe_ops.py]: I think these are the legacy ways of writing gradient checkpoint forward in old transformers, now they use a new class for enabling gradient checkpoint so that you don't need to write an ugly ifelse. But the meaning are similar, it's okay to put it here

File: src/lmms_engine/parallel/parallelize.py
Comment (kcz358) [inline src/lmms_engine/parallel/parallelize.py]: A bit ugly here to avoid the nested import. Is it possible to do it in the same way in qwen3_moe? I remember there is no circular import in qwen3_moe

File: src/lmms_engine/datasets/processor/__init__.py (hunk 1 above)
Comment (ngquangtrung57) [reply src/lmms_engine/datasets/processor/__init__.py]: No, there is a duplicate of "WanVideoDataProcessor"

File: src/lmms_engine/datasets/processor/qwen3_omni_moe_processor.py
[the remark this answers is not in the document]
Comment (ngquangtrung57) [reply src/lmms_engine/datasets/processor/qwen3_omni_moe_processor.py]: They are pretty similar actually, but some logics are different so i decide to keep them seperate

File: src/lmms_engine/models/qwen3_omni_moe/qwen3_omni_moe_ops.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        # EP is enabled - ParallelStyle._input_fn will handle the split
+        out_experts_split = self.experts(routed_input, num_tokens_per_expert)
+    else:
+        # EP is disabled, need to split routed_input manually
+        routed_input_split = torch.split(
+            routed_input,
+            split_size_or_sections=num_tokens_per_expert.tolist(),
+            dim=0,
+        )
+        out_experts_split = self.experts(*routed_input_split)
```
[the remark this answers is not in the document]
Comment (ngquangtrung57) [reply src/lmms_engine/models/qwen3_omni_moe/qwen3_omni_moe_ops.py]: When EP is disabled, there no _input_fn to split the input, so experts.forward() would receive the wrong argument (unsplit tensor + counts instead of split tensors per expert). I got matmul shape mistmatch without this. Not sure if samething happen with Qwen 3 MOE

**Review by kcz358 — APPROVED** [review]

**Comment by kcz358** [comment]
Thanks for the contribution. For future commit, feel free to config gpg or ssh key (I use this) so that your commit can be auto signed. I will bypass and merge this for now.
